lib/src/local_working_copy.rs

```diff
@@ -1218,6 +1212,44 @@ impl FileSnapshotter<'_> {
         Ok(())
     }
 
+    /// Emits file paths that don't exist in the `present_entries`.
+    fn emit_deleted_files(
+        &self,
+        dir: &RepoPath,
+        file_states: FileStates<'_>,
+        present_entries: &PresentDirEntries,
+    ) {
```

Comment:
issue: Does this work when using Watchman? In particular: I thought `file_states` was strictly an optimization, and we previously didn't rely on the presence or absence of a file as a canonical data source, but now it seems that we're checking `file_states` to determine if a file was deleted. But under Watchman, we don't populate `file_states`, so I'm worried.

Alternatives to sending all the files over a channel (I've tried these out at a previous workplace and they both resulted in speedups):
- Store all files in thread-local storage, then collect them after the full traversal is finished.
  - I think we already block on the full traversal before we can start processing the present files?
- Introduce a `BufferedSender`/`BufferedReceiver` wrapper around channels.
  - I suspect there's unnecessary contention when we try to send to many individual items over a channel, and we don't really need to optimize for the latency of individual items.

Reply:
It relied on `file_states.all()` before, so there should be no change conceptually.

BTW, I noticed `deleted: HashSet<_> = file_states.all()...` (in the original impl) has measurable cost in gecko-dev benchmark. I think it's better to move filtering to per-directory jobs, and reduce string length to be hashed (full path -> file name).

Reply:
I looked at my previous code and it seems like I misremembered, and it was always traversing `file_states.all()` even in the Watchman case, sorry. I'm guessing I tried to implement it without materializing the full `file_states` map, but didn't succeed/have time.

> BTW, I noticed deleted: HashSet<_> = file_states.all()... (in the original impl) has measurable cost in gecko-dev benchmark. I think it's better to move filtering to per-directory jobs, and reduce string length to be hashed (full path -> file name).

Agreed, but we might fundamentally struggle with the performance as long as we're always loading and deserializing O(repo) data from disk.

[just an idea] For engineering practicality, we could potentially say "for repos over a certain size, we expect you to use fsmonitor, and will optimize the fsmonitor snapshotting specifically" rather than trying to optimize both the non-fsmonitor and fsmonitor snapshotting cases.

[aside] For the Watchman case, I suspect I may have tried to load the previous tree for comparison rather than checking `file_states`. The idea would be that we only need to check O(change) entries instead of O(repo) entries. But it's possible that either I didn't really try it, or, if I tried it, then it was too slow to fetch too many arbitrary objects from disk, whereas even the O(repo) sequential loading of `file_states` is still faster than many random accesses.

Reply:
> we might fundamentally struggle with the performance as long as we're always loading and deserializing O(repo) data from disk.

We might need to redesign the tree state format (to be e.g. `mmap()`-able), but I don't have any particular idea in mind. It might be also good to use a committed tree object as you said.